File: core/src/main/java/com/cha/kim/wedding/core/invitation/wedding/Wedding.java
package com.cha.kim.wedding.core.invitation.wedding;

import com.cha.kim.wedding.core.account.Account;
import com.cha.kim.wedding.core.couple.Couple;
import com.cha.kim.wedding.core.family.Family;
import com.cha.kim.wedding.core.gallery.Gallery;
import com.cha.kim.wedding.core.gift.Gift;
import com.cha.kim.wedding.core.schedule.Schedule;
import com.cha.kim.wedding.core.share.Share;
import com.cha.kim.wedding.core.story.Story;
import com.cha.kim.wedding.core.visit.Visitor;
import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;

import javax.persistence.*;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

@Entity(name = "wedding")
@Getter
@Builder
@AllArgsConstructor
@NoArgsConstructor
@ToString
public class Wedding {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private int id;

    @Column(length = 255 , unique = true)
    private String url;

    @Column
    private boolean enable;

    @CreationTimestamp
    @Temporal(value = TemporalType.TIMESTAMP)
    @Column(updatable = false)
    private Date createDate;

    @CreationTimestamp
    @Temporal(value = TemporalType.TIMESTAMP)
    @Column(insertable = false)
    private Date updateDate;

    @Temporal(value = TemporalType.TIMESTAMP)
    @Column
    private Date weddingDate;

    @Column(length = 20)
    private String country;

    @Column(length = 20)
    private String city;

    @Column(length = 200)
    private String location;

    @Column
    private float latitude;

    @Column
    private float longitude;

    @Column(length = 255)
    private String hallImageUrl;

    @Column(length = 100, nullable = false)
    private String title;

    @Column(length = 1000 , nullable = false)
    private String introduction;
    @Column(length = 255)
    private String imageUrl;

    @Column(length = 1000)
    private String greetings;

    @Column(length = 1000)
    private String giftSmsGreetings;

    @OneToMany(mappedBy = "wedding" ,cascade = CascadeType.ALL , orphanRemoval = true)
    private List<Couple> couples = new ArrayList<>();

    @OneToMany(mappedBy = "wedding" ,cascade = CascadeType.ALL , orphanRemoval = true)
    private List<Family> families = new ArrayList<>();

    @OneToOne(mappedBy = "wedding" , cascade = CascadeType.ALL )
    private Gallery gallery = new Gallery();

    @OneToMany(mappedBy = "wedding" , cascade = CascadeType.ALL , orphanRemoval = true)
    private List<Schedule> schedules = new ArrayList<>();

    @OneToMany(mappedBy = "wedding" , cascade = CascadeType.ALL)
    private List<Share> shares = new ArrayList<>();

    @OneToMany(mappedBy = "wedding" , cascade = CascadeType.ALL , orphanRemoval = true)
    @OrderBy("orderIndex ASC")
    private List<Story> stories = new ArrayList<>();

    @OneToMany(mappedBy = "wedding" , cascade = CascadeType.ALL , orphanRemoval = true)
    private List<Gift> gifts = new ArrayList<>();

    @JsonIgnore
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "accountId" , referencedColumnName = "id")
    private Account account = new Account();

    @OneToMany(mappedBy = "wedding" , fetch = FetchType.LAZY ,cascade = CascadeType.ALL)
    private List<Visitor> visitors = new ArrayList<>();

    public void addGallery(Gallery gallery) {
        this.gallery = gallery;
    }

    public void addCouple(Couple couple) {
        this.couples.add(couple);
    }


    public void addSchedule(Schedule schedule) {
        this.schedules.add(schedule);
    }

    public void updateEnable(boolean enable) {
        this.enable = enable;
    }

    public void addStory(Story story) {
        this.stories.add(story);
    }

    public void addGift(Gift gift) {
        this.gifts.add(gift);
    }

    public void addFamily(Family family) {
        this.families.add(family);
    }
}

File: admin/src/main/webapp/resources/js/invitation/wedding/Wedding.js
var Wedding = function() {
    this.id = 0;
    this.dateTime = '';
    this.country = '';
    this.city = '';
    this.location = '';
    this.title = '';
    this.introduction = '';
    this.enable = false;
    this.couples = new Array();
    this.schedules = new Array();
    this.gallery = new Gallery();
    this.stories = new Array();
    this.imageUrl = '';
    this.latitude = '';
    this.longitude = '';
    this.hallImageUrl = '';
    this.families = new Array();
    this.gifts = new Array();
    this.url='';
    this.greetings='';
    this.giftSmsGreetings = '';
}

Wedding.createByForm = function ($form) {
    var wedding = new Wedding();
    wedding.id = $form.find('.wedding_id').val();
    wedding.dateTime = new Date($form.find('.dateTime').val());
    wedding.country = $form.find('.country').val();
    wedding.city = $form.find('.city').val();
    wedding.enable = $form.find('.enable').val();
    wedding.location = $form.find('.location').val();
    wedding.title = $form.find('.title').val();
    wedding.introduction = $form.find('.introduction').val();
    wedding.greetings = $form.find('#greetings').val();
    wedding.giftSmsGreetings = $form.find('#giftSmsGreetings').val();
    wedding.couples = createCouples($form);
    wedding.schedules = createSchedules($form);
    wedding.stories = createStories($form);
    wedding.gifts = createGifts($form);
    wedding.families = createFamilies($form);
    wedding.gallery = Gallery.createByForm($form.find('#gallery_item_area'));
    wedding.imageUrl = $form.find('.wedding_image_url').val();
    wedding.latitude = $form.find('.latitude').val();
    wedding.longitude = $form.find('.longitude').val();
    wedding.hallImageUrl = $form.find('.wedding_hall_image_url').val();
    wedding.url = $form.find('.wedding_url').val();

    function createCouples($form) {
        var couples = new Array();
        couples.push(Couple.createFemale($form));
        couples.push(Couple.createMale($form));
        return couples;
    }

    function createSchedules($form) {
        var schedules = new Array();

        $form.find('.schedule_item').each(function(index , item) {
            schedules.push(new Schedule($(item)));
        });

        return schedules;
    }


    function createGifts($form) {
        var gifts = new Array();

        $form.find('.gift_item').each(function(index , item) {
            gifts.push(new Gift($(item)));
        });

        return gifts;
    }

    function createFamilies($form) {
        var families = new Array();

        $form.find('.family_item').each(function(index , item) {
            families.push(new Family($(item)));
        });

        return families;
    }

    function createStories($form) {
        var stories = new Array();

        $form.find('.story_item').each(function(index , item) {
            stories.push(new Story($(item)));
        });

        return stories;
    }

    return wedding;
}

Wedding.createByUrl = function(url) {
    var wedding = new Wedding();
    wedding.url = url;
    return wedding;
}

Wedding.createByEnable = function(id , enable) {
    var wedding = new Wedding();
    wedding.id = id;
    wedding.enable = enable;
    return wedding;
}